lib/Dialect/Arc/Ops.cpp

```diff
@@ -135,6 +136,17 @@ LogicalResult StateOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
   return success();
 }
 
+LogicalResult StateOp::canonicalize(StateOp op, PatternRewriter &rewriter) {
+  // When there are no names attached, the state is not externaly observable.
+  // When there are also no internal users, we can remove it.
+  if (op->use_empty() && !op->hasAttr("name") && !op->hasAttr("names")) {
```

Comment:
Observability being determined by optional name attributes sounds a bit fragile  so it might be nicer to explicitly represent visibility/observability in the future. WDYT?

Reply:
Completely agree. We would like to redesign the `arc.state` op in general such that it doesn't handle that many things at once. In particular, I would like to remove the names attributes completely and merge them with the `arc.tap` construct. I would also like to split off the enable and reset operands and represent them differently. However, that still needs some discussions on how to do it exactly.

Reply:
`arc.tap` makes a ton of sense!

Reply:
This sounds very very nice! :+1: